<a href="https://colab.research.google.com/github/Pedro-Boechat/trainee/blob/main/periodo4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn
from sklearn.decomposition import PCA
import time

# Binário

In [227]:
df = pd.read_csv('/content/creditcard.csv')

In [228]:
df.head(5)

Time        V1        V2        V3  ...       V27       V28  Amount  Class
0   0.0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62      0
1   0.0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69      0
2   1.0 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66      0
3   1.0 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50      0
4   2.0 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99      0

[5 rows x 31 columns]

In [ ]:
df.isna().sum()

In [229]:
df = df.dropna()

In [230]:
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

In [231]:
df = df.drop(columns='Time')

In [232]:
colunas = df.columns

## Normalizar

In [207]:
for i in colunas:
  df[i] = (df[i] - df[i].min()) / (df[i].max() - df[i].min())

In [208]:
df.head()

V1        V2        V3        V4  ...       V27       V28    Amount  Class
0  0.935192  0.766490  0.881365  0.313023  ...  0.418976  0.312697  0.005824    0.0
1  0.978542  0.770067  0.840298  0.271796  ...  0.416345  0.313423  0.000105    0.0
2  0.935217  0.753118  0.868141  0.268766  ...  0.415489  0.311911  0.014739    0.0
3  0.941878  0.765304  0.868484  0.213661  ...  0.417669  0.314371  0.004807    0.0
4  0.938617  0.776520  0.864251  0.269796  ...  0.420561  0.317490  0.002724    0.0

[5 rows x 30 columns]

## Verificar balanceamento

In [233]:
from sklearn.utils import resample
# Separar a maioria e minoria
df_majority = df[df['Class']==0]
df_minority = df[df['Class']==1]
# Downsample
df_majority_downsampled = resample(df_majority, 
                                 replace=False,    
                                 n_samples=492) # Colocar aqui o downsample que quiser
# Combinar a maioria downsampled com a minoria
df = pd.concat([df_majority_downsampled, df_minority])

## Meu código

In [234]:
from sklearn.model_selection import train_test_split

X = df[colunas.drop('Class')]
Y = df['Class']
X_train, X_test, y_train, y_test = train_test_split(X,Y, test_size=0.2)


In [211]:
X_train.shape

(787, 29)

In [235]:
X_test = np.array(X_test)
X_train = np.array(X_train)
y_train = np.array(y_train)
y_test = np.array(y_test)

In [263]:
n_samples, n_features = X_train.shape
weights1 = np.random.rand(n_samples)
weights2 = np.random.rand(10)
hneurons = np.zeros(10)
bias1 = np.random.rand()
bias2 = np.random.rand()


In [237]:
def activation1(x): #Relu
  return x * (x > 0)

def dact1(x):
    return 1. * (x > 0)

def activation2(x): # ACTIVATION 2 é sigmoide
  return 1/(1 + np.exp(-x))


def dact2(x):
  return activation2(x) * (1.0 - activation2(x))


In [262]:
activation2(z2)

0.0008562779733391644

In [265]:
lista = []
predictions = []
iter = 200
learning_rate = 0.8
for _ in range(iter): # Para cada época
  predictions = []
  for i in range(len(X_train)): # Para cada X_train


    for k in range(len(hneurons)): # Para cada neuronio oculto
      lista = [] # Soma pesada de todos os neuronios de entrada
      for j in range(len(X_train[i])): # Para cada feature
        forward = (X_train[i][j] * weights1[j]) + bias1 # Pesa cada neuronio
        lista.append(forward) # Joga na lista da soma pesada

      z1 = np.sum(lista) # Z é a soma pesada
      ativada = activation2(z1) # Ativando Z com sigmoide
      hneurons[k] = ativada  # Cada neuronio oculto recebendo um valor
    soma = [] # Mesmo papel do "lista" anterior, mas agora da camada oculta pro output
    for p in range(len(hneurons)): # Para cada neuronio oculto
      forward = hneurons[p] * weights2[p] + bias2 # Pesa cada neuronio
      soma.append(forward) # Joga na lista da soma pesada
    z2 = np.sum(soma) # Faz a soma
    ativada_output = activation2(z2) # Ativa Z2 com a sigmoide
    if ativada_output >= 0.5: # Se o resultado for maior que 0,5: 1. Se nao, 0
      predicao = 1
    else:
      predicao = 0
  
    predictions.append(predicao) # Joga essa predicao numa lista com as predicoes dessa epoca

    cost = []

    custo = np.square(predicao - y_train[i]) # Custo
    cost.append(custo) # Essa linha e a proxima sao desnecessarias. Escrevi quando tava idiota mas nao muda nada
    custo_total = np.sum(cost)


### COMEÇANDO O BACKWARD. VERIFICAR SE OS CALCULOS DE DERIVADAS ESTAO CERTOS PQ EU NAO TENHO CTZ######


    for n in range(len(hneurons)): # Para cada neuronio oculto
      dw2 = (2 * (predicao - y_train[i]) * dact2(z2) * (hneurons[n])) # Derivada de cada um dos pesos 2
      weights2[n] = weights2[n] - learning_rate * dw2 # Atualizando um peso 2 de cada vez


    for g in range(len(X_train[i])): # Para cada neuronio de entrada
      dw1 = 2 * (predicao - y_train[i]) * dact2(z2) * sum(weights2) * dact2(z1) * (X_train[i][g]) # Derivada de cada um dos pesos 1
      weights1[g] = weights1[g] - learning_rate * dw1 # Atualizando um peso 1 de cada vez

    db2 = 2 *(predicao - y_train[i]) * dact2(z2) # Derivada do bias2
    db1 = 2 * (predicao - y_train[i]) * dact2(z2) * sum(weights2) * dact1(z1) # Derivada do bias 1


    bias1 -= learning_rate * db2 #Atualizando bias 1
    bias2 -= learning_rate * db2 #Atualizando bias2



In [261]:
hneurons[0] * weights2[0]

-1.238202150475155

In [260]:
hneurons[0]

1.0

In [259]:
activation2(z2)

0.0008562779733391644

In [258]:
X_train[0][0] * weights1[0]

-0.24155554695983902

In [268]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import plot_confusion_matrix

confusion_matrix(predictions, y_train)

array([[238, 224],
       [149, 176]])

In [24]:
1/(1 + np.exp(-z))

0.5

In [35]:
weights2

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [29]:
len(hneurons)

10

In [26]:
print(len(derw2))

830


In [27]:
print(sum(derw2))

5724.690429230827


## Daqui pra baixo são os códigos copiados ruins

In [ ]:
weights1 = []
weights2 = []
bias1 = 0
bias2 = 0
hidden = np.zeros(10)

def forward():
  for _ in range(iter):
    for idx, x_i in enumerate(X_train):
      forward = np.matmul(X_train,weights1) + bias1
      forward = activation1(forward)
      forward = np.matmul(forward, weights2) + bias2
      pred = activation2(forward)

      predictions = np.append(predictions, pred)
    
def backward():
  for i in range


Primeiro teste: Perceptron classification accuracy 0.9985955549313578

0.9982

array([[56831,    64],
       [   16,    51]])

0.76119

In [ ]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import plot_confusion_matrix

confusion_matrix(predictions, y_test)

array([[56831,    64],
       [   16,    51]])

In [ ]:
def relu(z):
    return x * (x > 0)
def forward_propagation(X, parameters):
    
    Z1 = np.dot(W1, X) + b1
    A1 = np.relu(Z1)
    Z2 = np.dot(W2, A1) + b2
    A2 = relu(Z2)
    cache = {"Z1": Z1,"A1": A1,"Z2": Z2,"A2": A2}
    
    return A2, cache
    

In [ ]:
def backward_propagation(parameters, cache, X, Y):
    #number of training example
    m = X.shape[1]
    
    W1 = parameters['W1']
    W2 = parameters['W2']
    A1 = cache['A1']
    A2 = cache['A2']
   
    dZ2 = A2-Y
    dW2 = (1/m) * np.dot(dZ2, A1.T)
    db2 = (1/m) * np.sum(dZ2, axis=1, keepdims=True)
    dZ1 = np.multiply(np.dot(W2.T, dZ2), 1 - np.power(A1, 2))
    dW1 = (1/m) * np.dot(dZ1, X.T) 
    db1 = (1/m)*np.sum(dZ1, axis=1, keepdims=True)
    
    grads = {"dW1": dW1, "db1": db1, "dW2": dW2,"db2": db2}
    
    return grads

In [ ]:
from scipy import stats

pred = np.array([])
k = 5
tam = len(distancias)
for i in range(tam):
  idxkneigh = distancias[i].argsort()[:k] # Pega os indices das k menores distancias
  vizinhos = [y_train[idxkneigh[j]] for j in range(k)] # Busca qual a classificação desses indices: red ou não
  mode_info = stats.mode(vizinhos) # Pega a moda das classificações dos KNN
  pred = np.append(pred, mode_info[0]) # Prediz

In [ ]:
acertos = 0
for i in range(tam):
  if pred[i] == y_test[i]:
    acertos += 1

print(acertos)

629


In [ ]:
acertos/(tam)

0.9828125

99% Precisão normalizado e com PCA

92% Precisão sem normalizar e sem PCA

98% Precisão normalizando e sem PCA